k2/csrc/pytorch_context.cu

```diff
@@ -149,18 +192,53 @@ class PytorchPinnedContext : public Context {
     return other.GetDeviceType() == kCpu;
   }
 
+  void CopyDataTo(size_t num_bytes, const void *src, ContextPtr dst_context,
+                  void *dst) override {
+    DeviceType device_type = dst_context->GetDeviceType();
+    switch (device_type) {
+      case kCpu:
+        // we assume that src and dst do not overlap
+        memcpy(dst, src, num_bytes);
+        break;
+      case kCuda: {
+        // from host to device
+        c10::cuda::CUDAStream stream = c10::cuda::getCurrentCUDAStream();
```

Comment:
just a question, @danpovey , which stream should we use here, the default stream or `dest->GetCudaStream()`?

Reply:
How are child streams created? I cannot find any code about child stream creation.
There are only several places in the documentation mentioning child streams.

Reply:
If we are going to use `dest->GetCudaStream()`, then we have to modify the interface `GetCudaStream()`.

https://github.com/k2-fsa/k2/blob/6854bbb8ff85c5da8b4aced629344299917489d2/k2/csrc/pytorch_context.cu#L205-L213

`THCCachingHostAllocator_recordEvent` requires an argument of type `CUDAStream`, which is a wrapper
around `cudaStream_t` with some other information, e.g., the priority of the stream and the device
associated with the stream. 

`CUDAStream` can be converted to `cudaStream_t` implictly, but the reverse is impossible.

I would propose to change `cudaStream_t GetCudaStream() const` to

```cpp
cudaStream_t GetCudaStream(void *pstream) const override {
   *(c10::cuda::CUDAStream*)pstream = c10::cuda::getCurrentCUDAStream();
   return *(c10::cuda::CUDAStream*)pstream;
}
```
The default value of `pstream` is nullptr, so the existing code will not be affected.

Reply:
> just a question, @danpovey , which stream should we use here, the default stream or dest->GetCudaStream()?


Fixed.

Reply:
I don't like this design because we call GetCudaStream a lot, and it can't be optimized to be inline because it's virtual.  I'd prefer to have a separate function return this extra info (e.g. GetTorchCudaStream() maybe.. we don't have to have this be defined for most instances of class Context.

Reply:
> I don't like this design because we call GetCudaStream a lot, and it can't be optimized to be inline because it's virtual.

The existing `GetCudaStream()` is virtual. If we add `GetTorchCudaStream()`, both of them
will be virtual and cannot be inlined.

I will change it to `GetTorchCudaStream()`.

Reply:
Yes it will still be virtual, but the point is, we don't want to add overhead to all the *other* calls to GetCudaStream().

There is still a slight problem here that we can only return the base Torch stream in this way, not any stream that might have been created by class ParallelRunner.

Reply:
Incidentally, if all this function call GetCudaStream() is doing is calling that Torch function, and since it can only be invoked from the Torch pinned-context object, there is no point using it at all-- better to just invoke that Torch function directly that gets the stream.

Reply:
You could perhaps find out whether there is a way to make Torch creaor provide additional streams when we need them, so we have this info.
Alternatively, we could ensure correctness by syncing the streams by means of events.  (Adds a bit of overhead, though).

Reply:
https://github.com/k2-fsa/k2/blob/0b055f07329cdca82d3f91c0b938803ba590644e/k2/csrc/pytorch_context.cu#L82-L85

The return value is from `g_stream_override.OverrideStream()`, which is its `stream_` that can be different from the one
returned from PyTorch.

Reply:
Yes, that is my concern, that for the streams which we created there is no way to convert them into Torch's notion of a stream.
It might be easier to always use our own allocator for pinned allocation, i.e. for the pinned allocators in the Torch and default cases to use the same underlying cached-allocator code (we would have to handle syncing by means of events).

Reply:
> It might be easier to always use our own allocator for pinned allocation

A PinnedContext of our own is added (not tested yet).